Grand Exchange Plugin

rbbi opened the pull request:
First implementation of Grand Exchange Plugin, some way to go on this - obviously!

TODO:
-  Fix progress bar size and location
-  Add a default icon for empty slots (not dwarf remains [id=0])!
-  Whatever else

![image](https://user-images.githubusercontent.com/31489752/32193992-f7aff128-bdb0-11e7-94d5-67847950e472.png)

![image](https://user-images.githubusercontent.com/31489752/32194075-34fcb4da-bdb1-11e7-8443-9751cf38f1b5.png)

UniquePassive commented:
First of all, good work, but please read about our [Coding Format](https://github.com/runelite/runelite/blob/master/CONTRIBUTING.md#-coding-format) (tabs, not spaces).

Please remove the 
```java
/**
 * @author Robbie, created on 29/10/2017 10:34 AM
 */
```
and instead put the license and copyright header at the top of new files.

Lastly, you missed [RSGrandExchangeOffer.java](https://github.com/runelite/runelite/pull/193/files#diff-30651bfd479780faf7e3738040d112d2) when fixing the format in your 2nd commit 😄 (braces on new lines).

That would probably be all to have your code looking good for when you've checked off your to-do.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java:
```diff
@@ -0,0 +1,625 @@
+
+  private void populate()
+  {
+    this.offerImages = new ImageIcon[8];
```
Consider using an array of objects for these instead of 6 arrays

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java:
```diff
@@ -0,0 +1,625 @@
+  private void handleLogic()
+  {
+    ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
+    ses.scheduleWithFixedDelay(() -> updateUI.accept(client.getGrandExchangeOffers()), 0, 1000, TimeUnit.MILLISECONDS);
```
There is an ```@Scheduled``` annotation to schedule periodic tasks. Although it may be better to hook to the offers field being modified in the client than to poll it.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java:
```diff
@@ -0,0 +1,625 @@
+  private String getIsBuySell(GrandExchangeOffer offer)
+  {
+    switch (offer.getProgress()) {
+      case 2:
```
These constants would be good to place into GrandExchangeOffer in runelite-api, maybe with some javadoc.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java:
```diff
@@ -0,0 +1,625 @@
+    panel_2.setBorder(new LineBorder(new Color(0, 0, 0)));
+    panel_2.setBackground(new Color(90, 82, 69));
+
+    imageLabels[1] = new JLabel("");
```
This to line 240 looks duplicated 8 times. Consider moving the shared logic to a method.

Adam- commented on runescape-api/src/main/java/net/runelite/rs/api/RSClient.java:
```diff
@@ -27,272 +27,272 @@
-	@Override
-	int getCameraX();
+public interface RSClient extends RSGameEngine, Client {
+  @Import("cameraX")
```
Please do not reformat large amounts of unrelated code.

rbbi commented:
Some progress on this plugin...
- JFrame-related component arrays are stored in their own respective object (GrandExchangeOfferSlot.java)
- Added Apache commons-lang dependency for ImmutablePair
- Rewrote core a bit

Still need to add various labels, fix progress bars, etc.

Adam- commented:
Thanks for your work on this so far. I don't like the use of Pairs (what is getLeft() and getRight()?) - and certainly don't want to pull in a whole new dependency just for it. Maybe place the GrandExchangeOffer in the GrandExchangeOfferSlot or vise versa. Or, if you must keep the pair-like structure make an object which holds both that is more descriptive.

Adam- commented:
@rbbi I have cleaned up the plugin to get the design in line with the rest of the plugins. Are you still working on the features outlined in your first comment?

rbbi commented:
@Adam- thanks for your commits, yes - I've been working my way through the list, also adding new stuff to it that could also prove useful.

Adam- commented:
Done in #698
